kit/src/main/java/com/oracle/javafx/scenebuilder/kit/library/user/LibraryFolderWatcher.java

```diff
@@ -161,98 +161,110 @@ private void runDiscovery() throws InterruptedException {
-                        	        }
-                        	    }
-
-                        	    exploreAndUpdateLibrary(jarsAndFolders);
-
-                        	    library.updateExplorationCount(library.getExplorationCount()+1);
-                        	}
-                        	finally {
-                        		library.setExploring(false);
-                        	}
-                        }
-                    } while (wk.reset());
-                } catch(IOException x) {
-                    Thread.sleep(1000 /* ms */);
+
+                            // We reconstruct a full set from scratch as soon as the
+                            // dirty flag is set.
+                            if (isDirty) {
+                                // First put the builtin items in the library
+                                library.setExploring(true);
+                                try {
+                                    library.setItems(BuiltinLibrary.getLibrary().getItems());
+
+                                    // Now attempts to add the maven jars
+                                    List<Path> currentMavenJars = library.getAdditionalJarPaths().get();
+
+                                    final Set<Path> fxmls = new HashSet<>();
+                                    fxmls.addAll(getAllFiles(FILE_TYPE.FXML));
+                            	    updateLibrary(fxmls);
+
+                                    final Set<Path> jarsAndFolders = new HashSet<>(currentMavenJars);
+                                    jarsAndFolders.addAll(getAllFiles(FILE_TYPE.JAR));
+
+                                    Set<Path> foldersMarkers = getAllFiles(FILE_TYPE.FOLDER_MARKER);
+                                    for (Path path : foldersMarkers) {
+                                        // open folders marker file: every line should be a single folder entry
+                                        // we scan the file and add the path to currentJarsOrFolders
+                                        List<Path> folderPaths = LibraryUtil.getFolderPaths(path);
+                                        for (Path f : folderPaths) {
+                                            jarsAndFolders.add(f);
+                                        }
+                                    }
+
+                            	    exploreAndUpdateLibrary(jarsAndFolders);
+
+                                    library.updateExplorationCount(library.getExplorationCount()+1);
+                                }
+                                finally {
+                                    library.setExploring(false);
+                                }
+                            }
+                        } while (wk.reset());
+                    } catch(IOException x) {
+                        Thread.sleep(1000 /* ms */);
+                    }
                 }
-            }
 
+            }
+        }
+        finally {
```

Comment:
Do you think it would make sense to move the `InterruptException` thrown from the method signature to a `catch` in the new  `try {} finally {}` ?
I don't think there is any difference, but it might help clarifying how the `while (true) {}` is terminated. It might also help adding a comment to that catch like:
```
// Ignored: Typically, when UserLibrary::stopWatching is invoked, an interruptionException is triggered to stop
// this watch service 
```
that would also help understanding why `finalize` is needed.

Reply:
Well, there is already a catch at line 102 that catches both runDiscovery() and runWatching().
In my opinion it is better to leave both methods with the throws InterruptedException as they are and have a single catch clause that handles both. Infact an interrupt could happen just before runWatching() during runDiscovery().
Having a single catch block is better even if in the future there will be another runSomething() following runWatching()..
It is ok to better clarify with a comment atone 102 and, I add, to log at info level the end of the service/thread.

Reply:
That makes sense. Just maybe include the comment I suggested to the finally clause? No need to log at info level though.

Reply:
Ok I'll add the comment. I'm out for vacation these days. I will be back next monday.